### PR TITLE
make communicate functions thread safe

### PR DESCRIPTION
**Pull request by gilesknap**
This PR fixes issue #13 

I'm afraid I don't have an easy way to write a unit test for this change.

But the system test using mcipc and the code in #13 does demonstrate that this works.

I could write a system test that runs minecraft in a container and tests parallel execution of commands if you are interested. I have such tests working well in mciwb.

### CODE REVIEW

File: rcon/battleye/client.py [hunk 1]
```diff
@@ -17,7 +18,7 @@
 
 __all__ = ['Client']
 
-
+lock = Lock()
```
Comment (conqp) [inline rcon/battleye/client.py:21]: Please write global constants in capital case. I.e. `LOCK`.

File: rcon/battleye/client.py [hunk 2]
```diff
@@ -55,10 +56,11 @@ def receive(self, max_length: int = 4096) -> Response:
 
     def communicate(self, request: Request) -> Response:
         """Send a request and receive a response."""
-        with self._socket.makefile('wb') as file:
-            file.write(bytes(request))
+        with lock:
```
Comment (conqp) [inline rcon/battleye/client.py:59]: Should really all instances of `Client` share the same lock?
Would it be better or worse, if each instance of `Client` would have its own lock?

File: rcon/source/client.py [hunk 3]
```diff
@@ -1,14 +1,16 @@
...
+from threading import Lock
 
 from rcon.client import BaseClient
 from rcon.exceptions import SessionTimeout, WrongPassword
 from rcon.source.proto import Packet, Type
 
-
 __all__ = ['Client']
 
+lock = Lock()
```
Comment (conqp) [inline rcon/source/client.py:12]: Capitalization - see above.

File: rcon/source/client.py [hunk 4]
```diff
@@ -31,8 +33,9 @@ def __init__(
 
     def communicate(self, packet: Packet) -> Packet:
         """Send and receive a packet."""
-        self.send(packet)
-        return self.read()
+        with lock:
```
Comment (conqp) [inline rcon/source/client.py:36]: Again: Global lock or per-instance lock?

File: rcon/battleye/client.py (hunk 2 above)
Comment (conqp) [reply rcon/battleye/client.py:59]: I vote for a per-instance lock.
E.g. why should clientA wait for clientB if it is contacting an entirely different server?

**Review by conqp — CHANGES_REQUESTED** [review]

**Review by conqp — COMMENTED** [review]
Furthermore, this can lead to pitfalls, since only `Client.communicate()` will be thread-safe.
Not so `Client.send()` or `Client.read()` or `Client.login()`.
I'd prefer if the caller would handle the thread safety, e.g. by providing a `threading.Lock()` around appropriate calls.

File: rcon/battleye/client.py (hunk 2 above)
Comment (gilesknap) [reply rcon/battleye/client.py:59]: good point!

**Comment by gilesknap** [comment]
Thanks for reviewing. 

I agree re locking in the caller so will do a PR to mcipc instead.

**Comment by gilesknap** [comment]
@conqp I have changed my mind.

We do need a global lock. The whole issue is if two different clients are simultaneously communicating. Each of their communicate calls needs to be atomic. We don't want to get a different client's response.

Inside of my 'exploding snowballs' code I use multiple threads and a separate client for each thread. I believe I need to have a separate client per thread because I can't use a socket across multiple threads.

I was able to get snowballs working OK with a global lock in MCIPC but this only works for multiple threads in a single process. By design I expect to have more than one process connecting to the same RCON (so multiple players can use interactive python together).

Thus I assert that we do need a global lock in RCON and only for communicate because it is the only non-atomic (with respect to socket calls) function.

**Comment by gilesknap** [comment]
OK that doen's make sense. since RCON will be different processes too :-)

I've decided I'm a bit confused by this. If I'm creating separate clients they should have separate sockets and thus not have interference. Maybe the issue lies elsewhere in MCIPC's socket creation?

**Comment by conqp** [comment]
Closing this, as consensus appears to be to let the caller worry about thread-safety.

**Pull request closed without merging**